binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java

```diff
@@ -48,124 +47,120 @@ abstract class SerdeResolverUtils {
+		Serdes.Long().getClass().getName(),
+		Serdes.Float().getClass().getName(),
+		Serdes.Double().getClass().getName(),
+		Serdes.ByteArray().getClass().getName(),
+		Serdes.ByteBuffer().getClass().getName(),
+		Serdes.Bytes().getClass().getName(),
+		Serdes.UUID().getClass().getName());
+		
 	/**
-	 *  Return the closest matching configured {@code Serde<?>} bean if one exists, or the specified
-	 *  {@code fallbackSerde}, or finally a standard default serde if no fallback specified.
-	 *
+	 *  Return the {@code Serde<?>} to use for the specified type using the following steps until a match is found:
+	 *  <p><ul>
+	 *    <li>the closest matching configured {@code Serde<?>} bean if one exists</li>
+	 *    <li>the Kafka Streams built-in serde if the target type is one of the built-in types exposed by Kafka Streams
+	 *    (Integer, Long, Short, Double, Float, byte[], UUID and String)
+	 *    <li>the fallback serde if specified and not one of the Kafka Streams exposed type serdes</li>
+	 *	  <li>the {@link JsonSerde} if the target type is not exactly {@code Object}</li>
+	 *	  <li>the fallback as the last resort</li>
+	 *  </ul>
 	 * @param context the application context
 	 * @param targetType the target type to find the serde for
-	 * @param fallbackSerde the fallback serde in case no matching serde bean found in the context
-	 * @return serde to use for the target type
+	 * @param fallbackSerde the serde to use when no type can be inferred
+	 * @return serde to use for the target type or {@code fallbackSerde} as outlined in the method description
 	 */
-	static Serde<?> resolveForType(ConfigurableApplicationContext context, ResolvableType targetType, Serde<?> fallbackSerde) {
+	static Serde<?> resolveForType(ConfigurableApplicationContext context, ResolvableType targetType, @Nullable Serde<?> fallbackSerde) {
 
-		List<Serde<?>> matchingSerdes = findMatchingSerdes(context, targetType);
-		if (!matchingSerdes.isEmpty()) {
-			return matchingSerdes.get(0);
-		}
+		Class<?> genericRawClazz = targetType.getRawClass();
 
-		// We don't attempt to find a matching Serde for type '?'
-		if (targetType.getRawClass() == null) {
-			return null;
+		// We don't attempt to find a matching Serde for type '?' - just return fallback
+		if (genericRawClazz == null) {
+			return fallbackSerde;
 		}
 
-		Serde<?> serde = null;
-		Class<?> genericRawClazz = targetType.getRawClass();
-		if (Integer.class.isAssignableFrom(genericRawClazz)) {
-			serde = Serdes.Integer();
-		}
-		else if (Long.class.isAssignableFrom(genericRawClazz)) {
-			serde = Serdes.Long();
-		}
-		else if (Short.class.isAssignableFrom(genericRawClazz)) {
-			serde = Serdes.Short();
-		}
-		else if (Double.class.isAssignableFrom(genericRawClazz)) {
-			serde = Serdes.Double();
-		}
-		else if (Float.class.isAssignableFrom(genericRawClazz)) {
-			serde = Serdes.Float();
+		List<String> matchingSerdes = SerdeResolverUtils.beanNamesForMatchingSerdes(context, targetType);
```

Comment:
Nit: don't need to point to itself if you call your own methods.

Reply:
It is definitely a style thing as some Sonar complain w/o reference to it (even when static). However, I had it w/o before and I have no strong preference. Updating...

Reply:
That sound suspicious: a couple lines below you use a `getStandardDefaultSerde()` without `SerdeResolverUtils.` prefix...